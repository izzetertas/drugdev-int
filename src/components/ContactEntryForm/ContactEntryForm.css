.contactEntryWrapper {
  display: flex;
  flex-direction: column;
  margin: 0 16px;
  padding: 16px;
  margin-top: 20;
  width: 100%;
  justify-items: center;
  align-items: center;
}

.contactEntryHeader {
  width: 95%;
  margin: 8px 16px;
  text-align: center;
  border: 1px solid #bbbebb;
  font-size: 16px;
}

.contactEntryContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 95%;
  border: 1px solid #bbbebb;
}

.buttonContainer {
  margin: 16px 0;
  text-align: left;
  padding-left: 8px;
}

button {
  margin: 4px !important;
}

.formItem { 
  margin: 16px;
}