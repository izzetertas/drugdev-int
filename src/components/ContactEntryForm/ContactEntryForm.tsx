import React, { useState } from 'react'
import uuid from 'uuid/v4'
import get from 'lodash/get'
import Button from '@material-ui/core/Button';
import { FormControl, Input, InputLabel } from '@material-ui/core';

import './ContactEntryForm.css'

export interface ContactInfo {
  id: string,
  name: string,
  email: string
}

export interface ContactEntryFormProps {
  contactInfo?: ContactInfo,
  onSubmit: (contactInfo: ContactInfo | null) => void,
}

export default class ContactEntryForm extends React.Component<ContactEntryFormProps> {
  state = {
    email: get(this.props, 'contactInfo.email', ''),
    name: get(this.props, 'contactInfo.name', '')
  }

  render() {
    return (
      <div className='contactEntryWrapper'>
        <div className='contactEntryHeader'>
          <h2>{get(this.props, 'contactInfo.id') ? 'Update ' : 'Add New '}Contact</h2>
        </div>
        <div className='contactEntryContainer'>
          <form
            onSubmit={e => {
              e.preventDefault();
              this.props.onSubmit({
                id: get(this.props, 'contactInfo.id', uuid()),
                name: this.state.name,
                email: this.state.email
              })
            }}
          >
            <div className='formItem'>
              <FormControl>
                <InputLabel htmlFor="name">Name</InputLabel>
                <Input
                  id="name"
                  aria-describedby="name"
                  onChange={(e) => { console.log(e.target.value); this.setState({ name: e.target.value}) }}
                  value={this.state.name}
                />
              </FormControl>
            </div>
            <div>
              <FormControl>
                <InputLabel htmlFor="email">Email address</InputLabel>
                <Input
                  id="email"
                  aria-describedby="email"
                  onChange={(e) => this.setState({email: e.target.value})}
                  value={this.state.email}
                />
              </FormControl>
            </div>
            <div className='buttonContainer'>
              <Button type='submit'
                variant="contained"
                color="primary"
                className='button'
              >
                {get(this.props, 'contactInfo.id') ? 'Update' : 'Add'}
              </Button>
              <Button
                variant="contained"
                color="secondary"
                className='button'
                onClick={() => this.props.onSubmit(null)}
              >
                Cancel
          </Button>
            </div>
          </form>
        </div>
      </div >
    )
  }
}
