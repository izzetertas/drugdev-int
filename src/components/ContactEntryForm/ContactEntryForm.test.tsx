import ContactEntryForm from './ContactEntryForm'
import { Input } from '@material-ui/core';
import React, { useState } from 'react'

describe('<ContactEntryForm />', () => {
  const minProps = {
    onSubmit: jest.fn()
  }
  const onSubmitMock = jest.fn()

  const props = {
    contactInfo: { id: 'sdfsdfÂ£$sAS1', name: 'Izzet', email: 'izzet@gmail.com' },
    onSubmit: onSubmitMock
  }

  describe('when new record is entered', () => {
    
    const wrapper = shallow(<ContactEntryForm {...minProps} />)
    it('should render without exploding', () => {
      expect(snapshot(wrapper)).toMatchSnapshot()
    })

    it('should assign correct header', () => {
      expect(wrapper.find('h2').text()).toBe('Add New Contact')
    });
  });
    
  describe('when an update record is come', () => {
    const wrapper = shallow(<ContactEntryForm {...props}/>)

    it('should render without exploding', () => {
      expect(snapshot(wrapper)).toMatchSnapshot()
    })
    
    it('should assign correct header', () => {
      expect(wrapper.find('h2').text()).toBe('Update Contact')
    });

    it('should assign default value correctly', () => {
      const input = wrapper.find(Input).first()
      expect(input.prop('value')).toBe(props.contactInfo.name)
    });
  });

  describe('When name field is changed', () => {
    it('should change state correctly', () => {
      const wrapper = shallow(<ContactEntryForm {...minProps}/>)
      
      const input = wrapper.find(Input).first()
      expect(input.prop('value')).toBe('')
      const expected = 'Sude'
      input.simulate("change", { target: { value: expected } });
      expect(wrapper.state('name')).toBe(expected)
    });
  });

  describe('When email field is changed', () => {
    it('should change state correctly', () => {
      const wrapper = shallow(<ContactEntryForm {...minProps}/>)
      
      const input = wrapper.find(Input).at(1)
      expect(input.prop('value')).toBe('')
      expect(input.prop('id')).toBe('email')
      const expected = 'sude@gmail.com'
      input.simulate("change", { target: { value: expected } });
      expect(wrapper.state('email')).toBe(expected)
    });
  });

  describe('when form is submitted', () => {
    it('should invoke onSubmit prop with correct params', () => {
      const wrapper = shallow(<ContactEntryForm {...props}/>)
      const expected = {
        id: props.contactInfo.id,
        name: 'Sude',
        email: 'newEmail@gmail.com'        
      }
      
      const nameInput = wrapper.find(Input).first()
      nameInput.simulate("change", { target: { value: expected.name } });

      const emailInput = wrapper.find(Input).at(1)
      emailInput.simulate("change", { target: { value: expected.email } });

      wrapper.find('form').simulate('submit', { preventDefault: jest.fn() })
      expect(onSubmitMock).toHaveBeenCalledWith(expected)
    });
  });
});
