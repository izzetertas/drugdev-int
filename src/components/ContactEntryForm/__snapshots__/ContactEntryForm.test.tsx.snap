// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactEntryForm /> when an update record is come should render without exploding 1`] = `
<div
  className="contactEntryWrapper"
>
  <div
    className="contactEntryHeader"
  >
    <h2>
      Update 
      Contact
    </h2>
  </div>
  <div
    className="contactEntryContainer"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="formItem"
      >
        <WithStyles(FormControl)>
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="name"
          >
            Name
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="name"
            id="name"
            onChange={[Function]}
            value="Izzet"
          />
        </WithStyles(FormControl)>
      </div>
      <div>
        <WithStyles(FormControl)>
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="email"
          >
            Email address
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="email"
            id="email"
            onChange={[Function]}
            value="izzet@gmail.com"
          />
        </WithStyles(FormControl)>
      </div>
      <div
        className="buttonContainer"
      >
        <WithStyles(Button)
          className="button"
          color="primary"
          type="submit"
          variant="contained"
        >
          Update
        </WithStyles(Button)>
        <WithStyles(Button)
          className="button"
          color="secondary"
          onClick={[Function]}
          variant="contained"
        >
          Cancel
        </WithStyles(Button)>
      </div>
    </form>
  </div>
</div>
`;

exports[`<ContactEntryForm /> when new record is entered should render without exploding 1`] = `
<div
  className="contactEntryWrapper"
>
  <div
    className="contactEntryHeader"
  >
    <h2>
      Add New 
      Contact
    </h2>
  </div>
  <div
    className="contactEntryContainer"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="formItem"
      >
        <WithStyles(FormControl)>
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="name"
          >
            Name
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="name"
            id="name"
            onChange={[Function]}
            value=""
          />
        </WithStyles(FormControl)>
      </div>
      <div>
        <WithStyles(FormControl)>
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="email"
          >
            Email address
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="email"
            id="email"
            onChange={[Function]}
            value=""
          />
        </WithStyles(FormControl)>
      </div>
      <div
        className="buttonContainer"
      >
        <WithStyles(Button)
          className="button"
          color="primary"
          type="submit"
          variant="contained"
        >
          Add
        </WithStyles(Button)>
        <WithStyles(Button)
          className="button"
          color="secondary"
          onClick={[Function]}
          variant="contained"
        >
          Cancel
        </WithStyles(Button)>
      </div>
    </form>
  </div>
</div>
`;
